doc/lightning-utxopsbt.7.md

```diff
@@ -0,0 +1,60 @@
+lightning-utxopsbt -- Command to populate PSBT inputs from given UTXOs
+================================================================
+
+SYNOPSIS
+--------
+
+**utxopsbt** *satoshi* *feerate* *startweight* *utxos* \[*reserve*\]
+
+DESCRIPTION
+-----------
+
+*utxopsbt* is a low-level RPC command which creates a PSBT using unreserved
+inputs in the wallet, optionally reserving them as well.
+
+It deliberately mirrors the parameters and output of
+lightning-fundpsbt(7) except instead of an optional *minconf*
+parameter to select unreserved outputs from the wallet, it takes a
+compulsory list of outputs to use.
+
+*utxos* must be an array of "txid:vout", each of which must be
+reserved or available: the total amount must be sufficient to pay for
```

Comment:
I do not see a use-case for allowing `reserved` state?  If a UTXO is reserved for a different spend, we should not allow spending it: xref how `fundchannel` today works, if you `fundchannel` in parallel with the same UTXO, one will error due to being unable to reserve.

Reply:
If you're attempting an RBF, you'll very likely be spending a still `reserved` UTXO. You could require a user to 'unreserve' it first but that seems like a roundabout way to achieve the same result.

Reply:
If you are attempting an RBF, you could just mutate the PSBT without messing with the reservations, right?  Like what we are planning with `fundchannel`/`multifundchannel`.  We get a partial transaction pre-`fundchannel_start`, then once we have the addresses we mutate that PSBT without changing reservations.  That works, right?

Reply:
Contrasting with `fundpsbt`, `fundpsbt` does not select UTXOs that are reserved.  So I do not think it is safe to allow selection of reserved UTXOs either.

Consider the case where we have a `fundchannel`.  The user invokes it twice with two different peers but with the same UTXO set.  To protect against this, `fundchannel` has to go check the reservations by checking by `listfunds`, but this runs into atomicity problems (it could get reserved between when it queries `listfunds` and it performs `utxopsbt` by the other `fundchannel` --- or other operations like `withdraw` or novel plugins --- running in parallel).  Then the two parallel `funchannel` processes proceed, but at `sendpsbt` time only one of those will confirm, and the other channel is negotiated with the peer and useless.  The user has to manually `fundchannel_cancel` the channel that did not get in.  Now maybe you can argue that the user should not do that, but if the user is running multiple plugins on top of their `lightningd` (and we should *want* them to do that, what would be the point of implementing and maintaining plugins if we did not want them to do that?) then such conflicts are harder to handle.

See, reservations are locks.  If you violate a lock, parallel operations will screw with you.  Basic multithreading.  So we should really be a lot more serious about respecting locks.  At least we do not block when a resource (UTXO) is currently locked by somebody else, but instead fail immediately, so that we can handle this case.

Reply:
Agreed it's great to alter the PSBT if you have it, but if you only have the tx, you want to get the PSBT, and `utxopsbt` will do that for you.

There's a good reason that reserved-only is the default: otherwise it's up to the caller to ensure that they are entitled to extend the reservation.  See last commit.

> Consider the case where we have a fundchannel. The user invokes it twice with two different peers but with the same UTXO set. To protect against this, fundchannel has to go check the reservations by checking by listfunds, but this runs into atomicity problems (it could get reserved between when it queries listfunds and it performs utxopsbt by the other fundchannel --- or other operations like withdraw or novel plugins --- running in parallel). Then the two parallel funchannel processes proceed, but at sendpsbt time only one of those will confirm, and the other channel is negotiated with the peer and useless. The user has to manually fundchannel_cancel the channel that did not get in.

No, I wrote this code (before I wrote utxopsbt): it did listfunds and then tried reservepsbt and then could try again if it failed.  However, this was quite complex, and utxopsbt made it much simpler.

> See, reservations are locks. If you violate a lock, parallel operations will screw with you. Basic multithreading. So we should really be a lot more serious about respecting locks. At least we do not block when a resource (UTXO) is currently locked by somebody else, but instead fail immediately, so that we can handle this case.

No, they are probabolistic datastructures.  Reservations are "likely to work" inputs for creating new transactions.

This leaves significant other uses, including deliberately conflicting transactions (such as dual funding where you may prefer to do that than to be DoS out of inputs), combining transactions, RBF and cancelling transactions by deliberately double-spending.